### ISSUE TITLE
Check DosTmrQueryTime usage

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#19]
Mozilla uses the `DosTmrQueryTime` API in NSPR (and in NSS) for getting high resolution tick intervals and such. There are claims that this function doesn't always work correctly with the new IOAPIC from Mensys — after some time it can start returning biased numbers which breaks correct interval measurements and may affect window repainting and other things depending on timers in Mozilla. There are reports that this happens after suspending and resuming the computer, for instance.

We will have to check this. The affected sources are:
- https://github.com/bitwiseworks/mozilla-os2/blob/master/nsprpub/pr/src/md/os2/os2inrval.c
- https://github.com/bitwiseworks/mozilla-os2/blob/master/nsprpub/pr/src/md/os2/os2rng.c
- https://github.com/bitwiseworks/mozilla-os2/blob/master/security/nss/lib/freebl/os2_rand.c

Note that for NSPR, there is an environment variable, NSPR_OS2_NO_HIRES_TIMER that disables this API and switches it back to the old good QSV_MS_COUNT mode.

The OS4 guys claim that they have fixed the `DosTmrQueryTime` issue in their `os4apic.psd` implementation. According to them, "in order to get precise data from PIT, you also need  to take the state of ISR IRR into account". There is also a test on their site regarding this issue in Mozilla: http://ru2.halfos.ru/core/downloads/kernel/os4/Test/timerTest.html.

**Comment by dspiatkowski** [comment]
Gentlemen, for what it's worth, this ACPI fix does not work. I have the latest version instaled here (3.22.06), but unless I run the NSPR_OS2_NO_HIRES_TIMER setting my CONFIG.SYS Firefox will inevitably start exhibiting the "high CPU spike" symptom, which I understand is related to this.

My hardware config is a multi-core (5 core) CPU, latest ACPI, running 17x Firefox but can try B3 of 24.x if you want. Happy to test for you....

Thanks!

**Comment by lerdmann** [comment]
Are you 100% positive that the "high CPU spike" goes away when you add NSPR_OS2_NO_HIRES_TIMER ? And is this specific to only some websites or is it true in general ?
Looking through the newsgroups there was another issue with high load (lots of open sockets) with Firefox on some websites. The "cure" was to modify "inetcfg.ini" to set mtudiscover to 1.

**Comment by ydario** [comment]
I just discovered I'm running firefox 24.8b3 with SET NSPR_OS2_NO_HIRES_TIMER=1 and ACPI 3.22.06, and I have the 100% cpu issue.
So I rebooted without SET NSPR_OS2_NO_HIRES_TIMER=1 and firefox started behaving uncorrectly: e.g. http://www.corriere.it has a timer to reload the page every minute or so, without SET NSPR_OS2_NO_HIRES_TIMER=1 the reload happens every few seconds. Also gmail web site started working badly, the chat window is not loading correctly.

**Comment by dspiatkowski** [comment]
Lars,
Yes, that is correct. I run with this setting turned ON, otherwise, I see the constant CPU spikes. To confirm this behaviour I disabled this setting, re-booted, opened up 6 separate Firefox windows, a total of 21 tabs, withing 3 hrs of Firefox up and running I am now seeing consistent CPU spikes, they are not constant 100% hits though. Rather I see these 10-20 sec long batches of activity where all 5 CPU cores appear to activate (per "CPU Monitor" app), in parallel to this TOP shows FIREFOX.EXE as the one showing highest CPU usage.

Eventually, if I leave this up and running within the next 24 hrs I get to a nearly constant CPU spike, where all cores are being utilized all the time. At this point in time I try to shut Firefox down as even on my 5-core SMP box the OS basically becomes non-responsive.

**Comment by SilvanScherrer** [comment]
@dazarewicz David could you eventually comment on this issue?

**Comment by dazarewicz** [comment]
What kind of comment would you like?  You seem to be discussing 2 apparently unrelated issues in this ticket. 1) the value returned by the QueryTime function. I implemented an SMP safe version of this function in ACPI PSD 3.22.06 and tested it and verified it was working as designed, and 2) the effect on FF of the NSPR_OS2_NO_HIRES_TIMER option which apparently changes the way FF works in a substantial way and in a way that is unrelated to the value returned by the QueryTime function.

Are you actually saying that the *value* returned by a timer is causing "CPU spikes"? Or are you saying that changing the NSPR_OS2_NO_HIRES_TIMER setting is causing "CPU spikes" and you are just assuming that it must be caused by the value returned by the timer function?  Has anyone actually checked what is really going on?

**Comment by dmik** [comment]
@dazarewicz David, I checked that several times already. The only thing NSPR_OS2_NO_HIRES_TIMER does in Firefox is forbids using `DosTmrQueryTime` in favor of old good `DosQuerySysInfo(QSV_MS_COUNT, QSV_MS_COUNT)`.

The problem with using `DosTmrQueryTime` is that it starts reporting inaccurate intervals after some period of Firefox activity. And at some point the bias becomes so big that it screws up Firefox timers. Which in turn screws up many things that depend on trimer triggered events. The JS code in http://ru2.halfos.ru/core/downloads/kernel/os4/Test/timerTest.html perfectly demonstrates that.

**Comment by dazarewicz** [comment]
Ok, fair enough.  However, I see two issues with this analysis.
1) In a preemptive, multitasking, multithreaded environment, there is no way to know what an "accurate interval" or an "inaccurate interval" is, and
2) Using the program that has the problem to test for the problem cannot produce credible results.

So therefore, to resolve #1 a solution would be to compare the value of the two timers at regular intervals, and to resolve #2 use a program that is not FF. The result is here: http://88watts.net/downloads/TmrTest.zip. Some deviation is expected due to interference from other threads and because the two timers cannot be read at *exactly* the same time.  Using this test, I do not see any issues.

**Comment by dazarewicz** [comment]
Oops, typo. The link should be http://88watts.net/download/TmrTest.zip

**Comment by dmik** [comment]
Okay, I rechecked it all once again and must state that I see no problems with high res timers any more. To test I removed `NSPR_OS2_NO_HIRES_TIMER` from CONFIG.SYS, run Firefox, `timerTest.html` and three news sites in other tabs in parallel + a tab with YouTube where the videos were constantly playing (tried both HTML5 and Flash). The maximum bias I noticed was 1645 ms (1645 ms more than the target 5000 ms). And after more than 300 timer runs (each one is 5000 ms) I only got 12 biases more than 50 ms. I think that may be considered as totally sufficient.

@ydario I also tried what you mention and all works fine here. http://www.corriere.it/ doesn't reload every few seconds. gmail also works. (This is in separate tabs **in addition** to the above). I have ACPI **3.22.6.0** here. Can you try to run `timerTest.html` from the first comment under heavy Firefox load and observe the results?

@dspiatkowski This is not about CPU spikes, this is about timer inconsistency. Can you, as well, try to run `timerTest.html` from the first comment under heavy Firefox load and observe the results?

**Comment by dazarewicz** [comment]
How are you measuring the "target 5000ms"? From the way your tests are running, it appears that it is simply a DosSleep(5000). I assume you already know this, but I feel it needs to be reiterated here.  When you give up the CPU using a DosSleep(5000) or similar function, you are not guaranteed to get control back in 5000ms. It will be *at least* 5000ms but can be longer. Sometimes it will be *much* longer depending on what other threads need to run and the priority of those threads. There is no way to know how long it will be. This is the reason that the timerTest.html test is not useful for testing the timer. The only thing it does is measure how long it took to return from the DosSleep().

**Comment by dazarewicz** [comment]
When you have a system "under heavy load", the probability that the time it takes to return from a DosSleep() will be longer than the time specified increases. If some of those threads that are creating the "heavy load" are marked time critical such as audio or video processing threads, that probability increases even more.

**Comment by lerdmann** [comment]
Does that mean that java function "setTimeout" uses DosSleep instead of DosTmrQueryTime ?

**Comment by dazarewicz** [comment]
@lerdmann It doesn't really matter. Either way it has to sleep while waiting so the result is the same.

**Comment by dmik** [comment]
@dazarewicz it matters. JS uses the timer functionality for that, not DosSleep. And the timer functionaliry is either provided by DosTmrQueryTime or DosQuerySysInfo (depending on the value of NSPR_OS2_NO_HIRES_TIMER.

**Comment by dazarewicz** [comment]
@dmik No it doesn't matter. 
I'm sure you do not just spin waiting for the timer:
  while (TimerValue < RequestedTimeout);
I'm sure you sleep while waiting (which you should):
  while (TimerValue < RequestedTimeout) sleep(n);
But even if you did just spin, the kernel won't let you have the CPU for an unlimited amount of time and will switch to a different thread at some point. The end result is exactly the same -- you don't really know how log it has been because you don't have exclusive use of the CPU.

**Comment by ydario** [comment]
just removed the variable from console window and started firefox beta3, issue with corriere.it still here. after a few seconds it loads http://www.corriere.it/index.shtml?refresh_ce

**Comment by dmik** [comment]
@dazarewicz `DosSleep` is not called — but something like `DosWaitEventSem` is (which perhaps eventually ends up in the same "sleep" kernel code as `DosSleep` though). But the JS test **dos not** directly result into this call. There is something like a timer thread that calculates elapsed time using either `DosTmrQueryTime` or `DosQuerySysInfo` and then calls `DosWaitEventSem` with an argument corresponding to the shortest interval we need to wait to fire the next timer. And if interval calculation is broken (reports values that are too far from the reality) we will get inappropriate intervals for `DosWaitEventSem ` and, as a result, inappropriate timer events. Imagine: when a new timer is added, the timer thread saves the current elapsed time value as the start point and asks the kernel to wait for the duration of the timer. Then, a new timer gets added before the first one fires off. The wait is interrupted, the thread gets the current elapsed time value and looks through the timers' start points to see which ones are expired (to fire them off). And if, say, we add a 5000 ms timer and then the thread is interrupted after just 2500 ms but `DosTmrQueryTime` reports that 5000 ms have already elapsed, this timer will be fired off through it was actually 2500 ms.

And surely I know that the value for `DosWaitEventSem` is a suggestion and that the kernel may hold the thread for a longer period. However, With previous ACPI versions, I got very *short* intervals which were more than two times less than the target value (e.g. 2203 ms instead of 5000 ms, see the third comment).

Anyway, if it's not `DosTmrQueryTime` how else would you explain that this JS code gives correct intervals when you use `DosQuerySysInfo` and may show complete madness when you use  `DosTmrQueryTime` (given that this is the **only** change in the Mozilla code)?

@ydario Still, works here. Did you try the JS test I asked you to? Maybe it somehow depends on hardware configuration.

**Comment by dazarewicz** [comment]
Well intervals that are radically shorter than they should be are not good, however longer intervals are to be expected. I understand that since ACPI 3.22.06 it is working OK now. No need to discuss older versions. Are you saying that you still have problems with the current version? If so, I will need a way to test for the problem that does not use FF.

**Comment by dmik** [comment]
I see no problems with the current version so far it but Yuri claims that he still has them. He however reported to me on IRC that the timer test works well for him so I now doubt that his problem is related to trimer issues at all. I asked him to test more (under heavy load etc).

@ydario you may also try the TmrTest.zip test from David (see the comments above) to see if that one gives accurate values for you.

**Comment by ydario** [comment]
I tried timertest.html while corriere.it is loading (acpi 3.22.07):

17. 5009
18. 5010
19. 4391
20. 5041
21. 3295
22. 5019
23. 565
24. 5004
25. 5023
26. 5489
27. 4572
28. 757
29. 5014
30. 879
31. 5006

I tried also a private ACPI build by Lars and this were even worse.

**Comment by ydario** [comment]
At the same time, tmrtest.exe reports Diff from 0 to 2, so it sounds correct.

**Comment by dazarewicz** [comment]
That means there is only a 0 to 2 ms difference between what the hires timer reports and the lo res timer reports. Something is obviously going on to produce the results that timertest.html shows. It is just unclear exactly what is going on.

**Comment by dazarewicz** [comment]
Just thinking about this a bit. The hires timer returns values that are many times larger than the ms counter (obviously). Below about 3.5 seconds, that math to convert the hires timer deltas to milliseconds fits into a 32 bit integer. Above about 3.5 seconds it will overflow. Is it possible you are getting math errors because of overflowing the size of your variables?

**Comment by dryeo** [comment]
On 03/10/15 10:27 PM, dazarewicz wrote:
> Just thinking about this a bit. The hires timer returns values that are
> many times larger than the ms counter (obviously). Below about 3.5
> seconds, that math to convert the hires timer deltas to milliseconds
> fits into a 32 bit integer. Above about 3.5 seconds it will overflow. Is
> it possible you are getting math errors because of overflowing the size
> of your variables?

Looks like it though there are shifts and masks involved, the code is 
here, 
https://github.com/bitwiseworks/mozilla-os2/blob/master/nsprpub/pr/src/md/os2/os2inrval.c#L46

**Comment by dazarewicz** [comment]
Interesting. So this code only uses bits 4-35 and throws away bits 0-3 and bits 36-63. This causes a bit of instability, but not too bad. This still means that the values returned are 74572 ticks/sec versus 1000 ticks/sec with the ms counter. Ok, so this is how the timer is queried, but what really matters is how this value is used (ie. what the math is that uses it).

**Comment by dazarewicz** [comment]
It appears that there are numerous functions that convert ticks to seconds/ms/us and seconds/ms/us to ticks, any of these may have overflows in the math and could be causing your problems. (I haven't looked.) All of those functions need to be checked for overflow conditions at the resolution you are using.

Alternatively, you could postulate that the code is overflowing by 1 bit and change _PR_MD_INTERVAL_INIT() to compare to a 16 bit value (like 0xffff) instead of a 17 bit value (currently 0x186a0 or 100000) and see what happens. This would reduce your precision by 1 bit.

Is there some reason you want to use the hires timer when the ms counter seems to work just fine?

**Comment by dspiatkowski** [comment]
@dmik @dspiatkowski This is not about CPU spikes, this is about timer inconsistency. Can you, as well, try to run timerTest.html from the first comment under heavy Firefox load and observe the results?

I am now running FF38.8.0 release. I still see the CPU spikes, but maybe not as pronounced as before?

I've ran the timer test and am attaching two output files which captured the results. While the FF use under each of the test was not the same, I made an attempt to try the same web-sites, same pages, etc. 

Overall, when not using the SET NSPR_OS2_NO_HIRES_TIMER=1 option I find that the timer test produces more (significantly) "error" results, and the timer values differ by a greater amount then otherwise.
![80coupe_front_profile](https://cloud.githubusercontent.com/assets/9026330/21829592/0ea6d60e-d765-11e6-9af4-c1ae66cfd4dd.jpg)
[timerTest_results_NO_HIRES_TIMER.txt](https://github.com/bitwiseworks/mozilla-os2/files/697668/timerTest_results_NO_HIRES_TIMER.txt)

[timerTest_results_default.txt](https://github.com/bitwiseworks/mozilla-os2/files/697669/timerTest_results_default.txt)

**Comment by dmik** [comment]
I'm testing it (*note* the new test location: http://gus.biysk.ru/os4/test/timerTest.html) with the timer thread fix from #257 and all-new AN ACPI (3.23.05) and friends and don't see any significant deviations in NSPR_OS2_NO_HIRES_TIMER unset mode. First 600 iterations give values like 5006, 5005, 5000, 5002, 5004 etc. After 600 iterations of the test there is some sort of degradation as the delta goes as high as 50-100 ms and then comes back below 50 ms. I wonder what it could be but that doesn't look fatal.

However, If I try to do the Gmail test (#248), I get weird delays which are more than the wait interval of 5000 ms itself:
```
782. 10918
783. 7709
784. 7754
785. 6610
786. 6955
787. 6689
788. 5090
```
It gets back to normal when I close Gmail. Something gets really stuck somewhere. I have a strong guess these weird delays come not from the low level hires timer, but are caused by PM being stuck. And there are no weird values which are below 5000 ms like it used to be.

A fresh test drop of XUL is here: http://rpm.netlabs.org/test/x3.7z

@ydario @dspiatkowski could you please test for weird delays that are significantly lower than 5000 ms under normal and heavy load? (significantly higher delays are ok for now, see above). Note to remove NSPR_OS2_NO_HIRES_TIMER from the environment.

**Comment by dmik** [comment]
Just had it run for 1000 cycles — not a single drop, all errors are within 100 ms, most are within 10 ms.

**Comment by LewisR** [comment]
@dmik you should know that the "latest" released AN ACPI is 3.23.07, with 3.23.08 in beta. 3.23.05 is considerably out of date (not that I expect there to be any differences against the later builds).

**Comment by DavidMcKenna** [comment]
No joy here::

0. 5002
1. 5002
2. 5000
3. 5000
4. 5002
5. 5000
6. 5000
7. 5004
8. 5000
9. 5000
10. 2672
11. 1505
12. 5002
13. 5000
14. 5002
15. 5000
16. 5000
17. 2758
18. 5000
19. 5002
20. 5002
21. 5002
22. 5002
23. 5002
24. 4045
25. 5002
26. 4117
27. 55
28. 5000

  and that was just sitting there doing nothing else.

**Comment by dmik** [comment]
@LewisR ok good to know but if it were RPM installable I'd already have it installed ;)

@DavidMcKenna do you have ArcaOS installed on that machine? Which version of ACPI?

**Comment by DavidMcKenna** [comment]
Yes, ArcaOS 5.0.1 with ACPI 3.23.07 and all latest drivers from AN. Intel i3-6100 with H110 chipset.

**Comment by dryeo** [comment]
Not sure if you've committed everything.
I rebuilt SeaMonkey after commit c98fa55 and while the browser seems 
smoother with NSPR_OS2_NO_HIRES_TIMER=1 and the numbers aren't bad until 
doing the gmail test when I get,
...
21. 5000
22. 5074
23. 39045
24. 32396
25. 5160
26. 5024
27. 5000
...
With NSPR_OS2_NO_HIRES_TIMER= and a light load, I got consistent 5002's 
while simply trying to load gmail in a tab gave me,
...
0. 5004
1. 5002
2. 5002
3. 5002
4. 5002
5. 5002
6. 5002
7. 5002
8. 5002
9. 5002
10. 5002
11. 5002
12. 5002
13. 5002
14. 5375
15. 974
16. 61
17. 265
18. 1008
19. 5656
20. 5416
21. 10124
22. 4875
23. 2597
24. 5182
25. 4336
26. 2703
27. 2615
28. 162
29. 326
30. 1766
31. 5002
...
and the tab wouldn't finish loading, with the something went wrong 
message and gmail reloading. At 31 I closed the tab. Earlier I left it 
open longer with no joy as the page never finished loading. It also took 
quite a while for the CPU load to drop back to close to zero (still 
higher then previous), after running at about 50%+25% across the cores. 
The numbers also dropped back to a pretty consistent 5004 with the odd 
low drop, one to 236.
This is on ArcaOS with the latest ACPI, acpi-3.23.08-20180118-test.

**Comment by dmik** [comment]
@DavidMcKenna what hardware? Did you try the OS/4 kernel by a chance? To see if you get the same there. Here both OS/2 and OS/4 behave exactly the same in all modes.

@dryeo no, all is there. Anyway, I think some kernel debugging is needed to understand it (beyond my knowledge, perhaps Steven or David from AN could assist here). The ~100x times dropdown is certainly a kernel level problem. Please also try OS/4 as a test.

**Comment by DavidMcKenna** [comment]
Hardware is a Gigabyte H110M-S2H motherboard with Intel i3-6100 3.7GHz processor with integrated video (HT turned off in BIOS), 8GB DDR4 RAM, 128GB SSD + 1TB rotating drives (SATA III, AHCI), USB wireless keyboard and mouse (board only has USB3, so added a USB 2 card), 1680x1200 monitor via DVI.

 I tried the OS/4 kernel and it worked fine, never saw a low value, occasionally high when playing youtube. I should mention that with ..NO_HIRES_TIMER=1 and AN kernel it works well too.

**Comment by dspiatkowski** [comment]
Hey Team!

On Wed, 31 Jan 2018 12:24:39 +0000 (UTC) Dmitriy Kuminov wrote:
> A fresh test drop of XUL is here: http://rpm.netlabs.org/test/x3.7z
>
> @ydario @dspiatkowski could you please test for weird delays that are  
> significantly lower than
> 5000 ms under normal and heavy load? (significantly higher delays are ok  
> for now, see above).
> Note to remove NSPR_OS2_NO_HIRES_TIMER from the environment.

So I ran a whole bunch of tests here (9 actually), all with a pretty decent  
amount of regular use over extended amount of time, pretty heavily loaded,  
about 10 windows open, multiple tabs/window.

I've attached a ZIPed results file along with a descriptive LOG, in short  
here are some key conclusions:

1) NSPR_OS2_NO_HIRES_TIMER is not usable here, similarly to what Dave  
reported, the timing from the test is all over the place, my normal use is  
to shut that OFF through the CONFIG.SYS switch set to 'SET  
NSPR_OS2_NO_HIRES_TIMER=1'

2) I left the previous release of libc066 in place for this test, as I have  
been running FF with the pair of libc066 and GC-fix XUL.DLL, so this time I  
only replaced the XUL.DLL

3) as hard as I try, I simply can not conclude much from the timings, they  
appear to be all over the place, some worse than others, by significant  
amounts

4) due to #3 incosistencies I re-ran all the tests with HPET in both ON and  
OFF modes to understand whether this has any impact, still inconclusive,  
although on my hardware running with HPET OFF appears to give better timing  
results

Finally, as I type this, my current configuration is such that the only  
time the CPU spikes are occuring is when I have focus on a FF window. This  
is weird, I have not seen this before. So literally while I am typing this  
email the CPU meter shows 0% use, as soon as I move the cursor over to a FF  
window the CPU spikes begin, the matching timerTest log is =>  
timerTest-TIMER_fix-HPET_off-HIRES_off-test9.html, and the timing results  
there look absolutely horrifying.

Sorrry, tried to nail this one down, did a pile of testing but got nowhere  
as best as I can tell.

My hardware configuration is: ACPI 3.23.07, W4.52 SMP mode here, Phenom 2  
X6 - 4GHz, 5 cores active.

  - Dariusz

**Comment by dryeo** [comment]
@dmik, testing with the os4kernel saw no numbers below 5000 with some 
high numbers, often correlating to high CPU usage. Only one core was 
enabled with the os4kernel+os4apic.psd.
I also tested with the usual kernel installed by ArcaOS and acpi.psd 
remmed out, this gave about the same numbers or perhaps slightly better 
as the os4kernel test, once again with one core enabled. Still higher 
numbers and CPU usage with NSPR_OS2_NO_HIRES_TIMER= compared to 
NSPR_OS2_NO_HIRES_TIMER=1.
The problem with low numbers seems to be in acpi.psd

**Comment by dmik** [comment]
@DavidMcKenna @dryeo what you say just proves the old thesis: somehow the new ACPI breaks the DosTmrQueryTime API. Then the recommendation to use NSPR_OS2_NO_HIRES_TIMER=1 should still be there. And as Firefox has nothing to do with this problem other than not use this API together with the current AN ACPI version, I will close this issue. Thanks for testing!

**Comment by dmik** [comment]
Just for the record, created an AN ticket for the issue (https://mantis.arcanoae.com/view.php?id=1702) as it's clearly related to the AN ACPI.PSD.

**Comment by dmik** [comment]
@dryeo @DavidMcKenna could you please run this test http://gus.biysk.ru/os4/test/tmrtest.exe on your machines where Firefox screws up? This test is a sole `DosTmrQueryTime` call with no FF logic. Just to see if it gives you any significant bias.

**Comment by DavidMcKenna** [comment]
When I run tmrtest.exe, it looks like:

rc=0 Resolution: 1193167
Elapsed 0 sec. Must be 0
Elapsed 2 sec. Must be 2

 and continues like that for 40 seconds, then disappears from the screen. All values are 2 seconds apart and match on every line.

**Comment by dspiatkowski** [comment]
On Mon, 12 Feb 2018 13:08:53 +0000 (UTC) DavidMcKenna wrote:
>
> When I run tmrtest.exe, it looks like:
>
> rc=0 Resolution: 1193167
> Elapsed 0 sec. Must be 0
> Elapsed 2 sec. Must be 2
>
> and continues like that for 40 seconds, then disappears from the screen.  
> All values are 2
> seconds apart and match on every line.

...same results here.

- Dariusz
  1980 Dodge Diplomat Coupe, 360-4bbl, 727, 3.91 SG motivated sleeper!

**Comment by dryeo** [comment]
Here I ran it about 10 times. One time it jumped from 24 secs to 27 secs 
and the must be number fell behind by 2.
